### PR TITLE
Less freezing time in multiprocessing with large arrays

### PR DESCRIPTION
**Pull request by weixuanfu2016**
## What does this PR do?

1. Change the backend of the parameter `max_eval_time_mins` that controls how many minutes TPOT has to optimize a single pipeline. The backend is based on [stopit module](https://github.com/glenfant/stopit)

2. Change the backend of multiprocessing to dask instead of joblib

3. Fix a old issue in Windows. Now TPOT allow Control+C during optimization process when n_job !=1 in Windows

4. A new unit test for timeout function

5. Add `check_X_y` for checking dataset format. 

6. Update documentation

## Where should the reviewer start?

[base.py](https://github.com/weixuanfu2016/tpot/blob/joblib_timeout/tpot/base.py#L769-L793)

## How should this PR be tested?

```
# coding: utf-8
from sklearn.datasets import make_classification
from tpot import TPOTClassifier
# make a huge dataset
X, y = make_classification(n_samples=50000, n_features=200,
                                    n_informative=20, n_redundant=20,
                                    n_classes=5, random_state=42)

# max_eval_time_mins=0.1 means 6 seconds limits for evaluating a single pipeline 
tpot = TPOTClassifier(generations=5, population_size=50, offspring_size=100, random_state=42, n_jobs=2, max_eval_time_mins=0.1, verbosity=3) 
tpot.fit(X, y)
```

## Any background context you want to provide?

[parallel processes freezing when matrices are too big](https://github.com/joblib/joblib/issues/138)
[Joblib hangs without crashing](https://github.com/joblib/joblib/issues/125)

## What are the relevant issues?

#436 #422 

[you can link directly to issues by entering # then the number of the issue, for example, #3 links to issue 3]

## Questions:

- Do the docs need to be updated?  Yes, docs are already updated in the PR.
- Does this PR add new (Python) dependencies? Yes, `stopit` and `dask`

### CODE REVIEW

**Comment by rhiever** [comment]
I think the best workaround (for now) is to make TPOT *not* use multiprocessing when `n_jobs=1`, and put warnings in the documentation that enabling multiprocessing (`n_jobs!=1`) may be slow and prone to freezing with very large datasets.

sklearn has this same problem, right? e.g. if you use `cross_val_score` with `n_jobs!=1` for a very large dataset, it will also be slow and/or freeze.

**Comment by weixuanfu2016** [comment]
Yep, sklean has a open issue for [this](https://github.com/scikit-learn/scikit-learn/issues/5115)

**Comment by weixuanfu2016** [comment]
Although I set TPOT not use joblib when n_jobs=1, I found that many sklean's classes and functions used the built-in joblib and then freeze happened again.

File: tpot/base.py [hunk 1]
```diff
@@ -771,24 +765,32 @@ def _evaluate_individuals(self, individuals, features, classes, sample_weight=No
...
+                        sklearn_pipeline,
+                        features,
+                        classes,
+                        self.cv,
+                        self.scoring_function,
+                        sample_weight,
+                        timeout=self.max_eval_time_seconds
+                    )
+                    jobs.append(job)
+                tmp_scores = compute(*jobs, get=multiprocessing.get, num_workers=self.n_jobs)
```
Comment (mrocklin) [inline tpot/base.py]: There might be some benefit to checking in with Dask's globals before hard-wiring `dask.multiprocessing.get`.  You might try the following instead:

```python
import dask
if njobs == 1:
    get = dask.async.get_sync
elif 'get' in dask.context._globals:
    get = dask.contest._globals['get']
else:
    get = multiprocessing.get

compute(..., get=get)
```

This would allow users to step in with other schedulers, like the distributed scheduler.  Also if you are only using a single core then it's best to avoid the multiprocessing scheduler, which has some non-trivial data movement overhead.

**Comment by mrocklin** [comment]
Also, apologies for going off topic, but if you all ever wanted to dive more deeply into Dask you might consider the asynchronous concurrent.futures interface, notably [submit](http://distributed.readthedocs.io/en/latest/api.html#distributed.client.Client.submit) and [as_completed](http://distributed.readthedocs.io/en/latest/api.html#distributed.client.as_completed).  Here is an example of a somewhat-related library that uses this interface well to solve a somewhat-related computational problem: https://github.com/eriknw/dask-patternsearch .

Sorry for the intrusion.  I ran across this PR while searching github.

File: tpot/base.py (hunk 1 above)
Comment (weixuanfu2016) [inline tpot/base.py]: Thanks for you suggestions.

**Comment by weixuanfu2016** [comment]
@mrocklin No apology needed. I am very happy to find better solution. Thanks again for these good suggestions. Right now, the main issue is that the freezing with a large dataset seems from the scikit-learn and its joblib as mentioned here and #436 , which is very hard to solve until sklearn fixed it.  We used a thread-based timeout decorator for trying to kill these freezing jobs but somehow it is not works very well. The decorator can kill the freezing jobs but not under the right time limit. I suspect it is due to GIL. I will look into the example to check if it change solve our issues, Let me know if you have any idea about it.

**Comment by rhiever** [comment]
Looks like this PR has merge conflicts now too.

**Comment by weixuanfu2016** [comment]
OK, I will resolve these conflicts in `tests.py`

File: tpot/base.py [hunk 2]
```diff
@@ -449,6 +444,16 @@ def fit(self, features, classes, sample_weight=None):
                     'a more reasonable outcome from optimization process in TPOT.'
                     )
 
+        if (features.shape[0] > 10000 or features.shape[1] > 100) and self.n_jobs !=1:
```
Comment (rhiever) [inline tpot/base.py]: Warnings should only print if verbosity>2.

Comment (weixuanfu2016) [inline tpot/base.py]: Make sense. I fixed it.

**Comment by rhiever** [comment]
Looks like something broke with the rest of the merges. :-(

**Comment by weixuanfu2016** [comment]
Conflicts fixed

**Comment by weixuanfu2016** [comment]
I also add the patches in master branch (0.7.5) in this PR
